src/Sparrow/Allocators.Pool.FixedSize.cs

```diff
@@ -0,0 +1,245 @@
+            // that allows us to release some steam at the cost of hitting cold code. 
+            // https://github.com/dotnet/coreclr/issues/6024
+
+            allocator.InUse -= ptr.Size;
+            allocator.Allocated -= ptr.Size;
+
+            Pointer nakedPtr = new Pointer(ptr.Ptr, ptr.Size);
+            allocator._internalAllocator.Release(ref nakedPtr);
+        }
+
+        private void ThrowMemoryDoesNotBelongToAllocator(in Pointer ptr)
+        {
+            throw new InvalidOperationException($"The memory pointer {ptr.Describe()} does not belong to this allocator.");
+        }
+
+        private void ThrowOnlyFixedSizeMemoryCanBeRequested()
+        {
+            throw new InvalidOperationException($"The memory size requested is not supported by this allocator. You should use {this._options.MaxBlockSize} instead.");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Renew(ref FixedSizePoolAllocator<TOptions> allocator)
+        {
+            if (Allocator.LowMemoryFlag.IsRaised())
+            {
+                ReleaseMemoryPool(ref allocator);
+                allocator.Allocated = 0;
+            }                
+
+            allocator._internalAllocator.Renew();
+            allocator.InUse = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Reset(ref FixedSizePoolAllocator<TOptions> allocator)
+        {
+            if (allocator._options.HasOwnership)
+                ReleaseMemoryPool(ref allocator);
+            else
+                ResetMemoryPool(ref allocator);
+
+            allocator._internalAllocator.Reset();
+
+            allocator.TotalAllocated = 0;
+            allocator.Allocated = 0;
+            allocator.InUse = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void OnAllocate(ref FixedSizePoolAllocator<TOptions> allocator, Pointer ptr)
+        {
+            // Nothing to do here.
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void OnRelease(ref FixedSizePoolAllocator<TOptions> allocator, Pointer ptr)
+        {
+            // Nothing to do here.
+        }
+
+        public void Dispose(ref FixedSizePoolAllocator<TOptions> allocator)
+        {
+            if (allocator._options.HasOwnership)
+            {
+                // We are going to be disposed, we then release all holded memory. 
+                allocator.ReleaseMemoryPool(ref allocator);
+            }
+
+            allocator._internalAllocator.Dispose();
+        }
+
+        private void ResetMemoryPool(ref FixedSizePoolAllocator<TOptions> allocator)
+        {
+            // We dont own the memory pool, so we just reset the state and let the owner give us memory again on the next cycle.
+            // This is the typical mode of operation when the underlying allocator is able to reuse memory (ex. NativeAllocator).
```

Comment:
This comment and the one on `ReleaseMemoryPool` both mention `NativeAllocator`, so this is confusing.
What is the use cases for the pool not owning their memory?

Reply:
Composition, the idea is that the pool does not allocate memory, it uses a real allocator to do so. However, given that you can have Self Healing and Leaky allocators, the ownership is about transferring the responsibility. Pools don't need to release explicitly if the allocator is self-healing.

Reply:
There was an error in the comment. In the standard pool it was tight.